[DC-709] Retraction to sandbox and remove deactivated ehr participants.

amcgrenera-vumc opened the pull request.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+                        help='Ticket number to append to sandbox table names',
+                        required=True)
+    parser.add_argument(
+        '-pp',
+        '--pids_project_id',
+        action='store',
+        dest='pids_project_id',
+        help='Identifies the project where the pids table is stored',
+        required=True)
+    parser.add_argument(
+        '-pd',
+        '--pids_dataset_id',
+        action='store',
+        dest='pids_dataset_id',
+        help='Identifies the dataset where the pids table is stored',
+        required=True)
+    parser.add_argument('-pt',
```
This is a little odd.  The short form of a command is typically `-` followed by a single alphabetic character.  This allows chaining multiple single character short form commands into a single long chain, for example `-s`, `-t`, `-u` could be invoked from the command line as `-stu`.  I tested your argument parsing code and it works, but it isn't entirely conventional.  When I think about conventional short form options I think about common linux/macOS command options.  You can view examples of what I consider common command line conventions by using `man ls`, `man cat`, or `man tail`.  Use `q` to exit a man page.

This link references the common short form and long form of command line options.
https://docs.python.org/3/library/argparse.html#name-or-flags

One other thing is I might not mix the underscore and hyphen in the command line parameters.  We haven't set a precedent for this in our project yet, but I think `--pids-project-id` is a little cleaner than mixing them like `--pids_project_id`.

amcgrenera-vumc replied:
thanks for the reference link!

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+    parser = argparse.ArgumentParser(
+        description=
+        'Runs retraction of deactivated EHR participants on all datasets in project. The rows to be retracted'
+        'are based on the date entered and will be retracted if the date entered is before the retraction date'
```
nit:  `date.  '`  Right now, this is creating a run on sentence.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
```
Another suggestion, you can do some additional refactoring here.  You can write a function to parse all your arguments and return your parsed arguments.  Then you could have
```
def parse_args(raw_args=None):
    # add your existing parser definition code here
    return parser.parse_args(raw_args)


def main(args=None):
    args = parse_arguments(args)
    query_list = create_queries(args.project_id, args.ticket_number,
                                args.pids_project_id, args.pids_dataset_id,
                                args.pids_table)
    run_queries(query_list)
    logging.info('Retraction complete')

if __name__ == '__main__':
    main()
```

You may want to consider this approach because this will allow the module to be run the same from the command line or imported and used from another python module.

amcgrenera-vumc replied:
thank you for this suggestion-- much cleaner and easier to read. also glad it will make it possible to run the same for cmd line and imported.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+                                args.pids_project_id, args.pids_dataset_id,
+                                args.pids_table)
+    run_queries(query_list)
+    logging.info('Retraction complete')
```
We need to address the logging.  You are importing the logging module, but you're not setting it up.  Because we don't have a logging module setup for the project, you must set up the logging yourself to use it (log file location(s), date formatting, message formatting, any rotating handlers) .  If you comment these lines:
```
#    query_list = create_queries(args.project_id, args.ticket_number,
#            args.pids_project_id, args.pids_dataset_id,
#            args.pids_table)
#    run_queries(query_list)
    logging.info('Retraction complete')
```
I believe you are expecting to see `Retraction complete` on your console but nothing is printed.  The deid runner sets up a logging location and console logger.  The function I'm thinking of is here:  https://github.com/all-of-us/curation/blob/develop/data_steward/tools/run_deid.py#L40 .  

Additionally, to identify your module's logger you should add this line:
```
LOGGER = logging.getLogger(__name__)
```

Then instead of calling `logging.info()` you would call `LOGGER.info()` because it has the logging instance for your module.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+
+    date_columns_df = get_date_info_for_pids_tables(project_id)
+    logging.info(
+        'Dataframe creation complete to loop through to create retraction queries'
```
nit:  This logging statement is confusing.  I understand `Dataframe creation complete`.  The rest seems  like an incomplete thought.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+        'Dataframe creation complete to loop through to create retraction queries'
+    )
+
+    # Loop through the deactivated PIDS df to create queries based on the retractions needed per PID table
+    logging.info(
+        'looping through deactivated pids df to create queries based on teh retractions needed per pid table'
+    )
```
nit:  spelling.  `teh`.  because the comment and the log statement are almost exactly the same, the comment might not be necessary.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+    logging.info(
+        'looping through deactivated pids df to create queries based on teh retractions needed per pid table'
+    )
+    for ehr_row in deactivated_ehr_pids_df.itertuples(index=False):
```
I did not know about this pandas method.  Thanks!

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+        for date_row in date_columns_df.itertuples(index=False):
+            # Determine if dataset is deid to correctly pull pid or research_id and check if ID exists in dataset or if
+            # already retracted
+            if is_deid_dataset(date_row.dataset_id) is True:
```
Is this helper function really needed?  This could be rewritten as:
```
if re.match(DEID_REGEX, date_row.dataset_id):
```
If a match is returned, it is automatically True.  If no match exists, it is automatically false.  What is your design rationale?  Are you planning to use this somewhere else?

amcgrenera-vumc replied:
I like your way better as well. I initially thought I may need that helper function more than once, but now that I see I don't, it's unnecessary and clutters, thanks!

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+
+    research_id_df = bq.query(
+        RESEARCH_ID_QUERY.format(project=project, prefix_regex=prefix, pid=pid))
+    if research_id_df.empty:
```
Is this condition expected to happen?  Does it indicate an error condition?  Just curious.

amcgrenera-vumc replied:
just queried all `_deid_map` tables to see if there are any instances, did not find any so I will remove this.

amcgrenera-vumc replied:
I remember now why I added this-- while there will always be a `research_id` per `person_id` the record may have already been removed from the deid table depending on other cleaning rules.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+
+    research_id_df = bq.query(
+        RESEARCH_ID_QUERY.format(project=project, prefix_regex=prefix, pid=pid))
+    if research_id_df.empty:
+        logging.info('no research_id associated with person_id: %s ' % pid)
+        return None
+    else:
+        return research_id_df.to_string()
```
You could also write this as:
```
research_id_df.empty:
    logging.info(f'no research_id associated with person_id: {pid}')
    return ''   # or return None

return research_id_df.to_string()
```
The else statement here is unnecessary because it will execute the return even without the `else`.  Consider returning an empty string because it will act the same as `None` in conditional statements.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+    :param date_row:
+    :return:
+    """
+    check_pid_df = bq.query(
```
Check your pid before using it.  If the pid is either `None` or an empty string, `''`,  the formatted string will still be handled correctly by python but will cause a BigQuery failure.  If, for some reason, `get_research_id` returned `None` or an empty string both of these queries could be built by this function.  Both would fail.  You can check for this and return `False` without even querying BigQuery.  Or you can query BigQuery and catch the exception.  Then use the catch to set your return value.
```
SELECT
COUNT(*) AS count
FROM `aou-res-curation-test.lrwb_deid_test_input_20190910.observation`
WHERE person_id = None

BigQuery returns this error response:  google.api_core.exceptions.BadRequest: 400 Unrecognized name: None at [5:19]


SELECT
COUNT(*) AS count
FROM `aou-res-curation-test.lrwb_deid_test_input_20190910.observation`
WHERE person_id =

BigQuery returns this error response:  google.api_core.exceptions.BadRequest: 400 Syntax error: Unexpected end of script at [5:18]
```

amcgrenera-vumc replied:
I have adjusted where if the `research_id` results in None it will not run `check_pid_exist` and exit the loop.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+            # Check if PID is in table
+            if check_pid_exist(pid, date_row) is True:
+                # Get or create sandbox dataset
+                sandbox_dataset = get_sandbox_dataset_id(date_row.dataset_id)
+                datasets_obj = bq.list_datasets(project_id)
+                datasets = [d.dataset_id for d in datasets_obj]
+                if sandbox_dataset not in datasets:
+                    logging.info('%s dataset does not exist, creating now' %
+                                 sandbox_dataset)
+                    create_sandbox_dataset(date_row.project_id,
+                                           date_row.dataset_id)
+
```
I know you and Nishanth were wanting to have a `get_sandbox_dataset` function that guaranteed the sandbox dataset was created.  I was looking at the existing `create_sandbox_dataset` function and realized the reason you were having issues with using this function is because it has a bug.  Lines 20-24 should actually be:
```
def create_sandbox_dataset(project_id, dataset_id, raise_error=True):
    # Lines 16 - 19
    try:
        create_dataset(project_id=project_id,
                       dataset_id=sandbox_dataset_id,
                       friendly_name=friendly_name,
                       description=description,
                       overwrite_existing=False)
    except HttpError as err:
        if err.code != 409:
            raise
        elif raise_error:
            raise

    return sandbox_dataset_id
```
This allows `create_sandbox_dataset` to create a dataset if it doesn't exist (which is what it was trying and failing to do).  It also allows you to specify if you are just trying to ensure the dataset exists or if you want to raise the error.  The default is set to raise the error if the dataset already exists because that is how it was originally written.

You could make this change and replace lines 271-279 with:
```
    create_sandbox_dataset(date_row.project_id, date_row.dataset_id, raise_error=False)
```

amcgrenera-vumc replied:
For now, since there is a ticket to fix `sandbox.py`, I have updated this to the new function that was merged last week `check_and_create_sandbox_dataset`.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+            else:
+                # break out of loop to create query, if pid does not exist in table
+                continue
+    print(
+        'Query list complete, retracting ehr deactivated PIDS from the following datasets: %s'
+        % date_columns_df['dataset_id'].tolist())
```
did you intend to remove this line?

amcgrenera-vumc replied:
I sure did :)

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+    """
+    query_job_ids = []
+    for query_dict in queries:
+        if 'sandbox' in query_dict['destination_dataset_id']:
```
Right now, this business logic will work.  But, if we ever changed the naming convention of the sandbox datasets this condition could be impacted.  A suggestion to think about, would be to add a key/value pair to your query dictionary to indicate what type of query you are running.  You should be able to easily add this when you are generating your query.  It could be something like `{'type': 'sandbox' or 'retraction'}`.  Then you could check this and it would be guaranteed to work even if dataset names change.  Thoughts/concerns/suggestions?

amcgrenera-vumc replied:
per our conversation, totally agree. Adding a key/value to query then will check `if query_dict['type'] == 'sandbox':`

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+from sandbox import get_sandbox_dataset_id, create_sandbox_dataset
+from constants import bq_utils as bq_consts
+
+DEACTIVATED_PIDS_QUERY = """
```
I was recently introduce to the jinja templates.  I use one here:  https://github.com/all-of-us/curation/blob/develop/tests/integration_tests/data_steward/cdr_cleaner/cleaning_rules/bigquery_tests_base.py#L44

It allows us to format the sql nicely in code without sending a lot of useless blank characters through to the API.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+import pandas as pd
+
+# Project imports
+from utils import bq
+import bq_utils
```
Why are you using both the new `bq` module and the old `bq_utils` module?  We should prefer the new `bq` module.

amcgrenera-vumc replied:
There are a couple of functions that are in `bq_utils` that do not exist yet in `bq`, I have them listed below. I will also need to adjust or add in an additional function to return job results. Currently `bq.query` returns a dataframe, while `bq_utils.query` returns `bq_service.jobs()`. 

- `bq_utils.wait_on_jobs`
- `bq_utils.BigQueryJobWaitError`

Should these be additional tickets, or should I just make these adjustments/additions to `bq` in this PR?

lrwb-aou replied:
I think those functions will be dropped entirely from the `utils/bq.py` module.  You can do the same thing with the existing client library functionality.
```
client = bq.get_client(project_id)
response = self.client.query('Select count(*) from `project.dataset.observation`')

# wait for the result using client library functions
results = response.result()

# check for an exception
exc = response.exception()
if exc:
    raise exc
```
This will wait on the job to finish automatically.  You can raise the exception if the job didn't finish correctly.  Here are some links to help.

https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client.query

https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.QueryJob.html#google.cloud.bigquery.job.QueryJob.result

amcgrenera-vumc replied:
great, thank you for this, i was able to adjust!

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+    """
+    Get list of tables that have a person_id column; along with a separate
+    list of the datasets containing pids
+    """
```
nit:  Finish adding parameter and return info to the docstring.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+    :return: filtered dataframe which includes the following columns for each table in each dataset with a person_id
+    'project_id', 'dataset_id', 'table', 'date_column', 'start_date_column', 'end_date_column'
+    """
+    logging.info('getting all tables with person id in %s' % project_id)
```
I think the consensus from the last engineer sync meeting was to use f-strings as part of logging.  Please make this change.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,441 @@
+    'project_id', 'dataset_id', 'table', 'date_column', 'start_date_column', 'end_date_column'
+    """
+    logging.info('getting all tables with person id in %s' % project_id)
+    pids_tables, pids_datasets = get_pids_datasets_and_tables(project_id)
+    # Create empty df to append to for final output
+    date_fields_info_df = pd.DataFrame()
+
+    # Loop through datasets
+    logging.info(
+        'Looping through datasets to filter and create dataframe with correct date field to determine retraction'
+    )
+    for dataset in pids_datasets:
+        # Get table info
+        table_info_df = bq.get_table_info_for_dataset(project_id, dataset)
+        # Filter out to only records with pids
+        pids_tables_df = table_info_df[table_info_df['table_name'].isin(
+            pids_tables)]
+
```
I think this can be refactored some.  Right now, `get_pids_datasets_and_tables` uses this query:
```
SELECT table_name
FROM `{project}.{dataset}.INFORMATION_SCHEMA.COLUMNS`
WHERE COLUMN_NAME = 'person_id'
```
Then `get_table_info_for_dataset` is using this query:
```
SELECT table_name
FROM `{project}.{dataset}.INFORMATION_SCHEMA.COLUMNS`
```
These are nearly identical BigQuery calls.  You can probably use the results from this second query to populate your lists.  But I would argue that the `pids_tables` list is useless because you are only using it to get the list of tables with a `person_id` and then subselect the tables with a `'DATE'` field.

On the test project, this results in  879 bigquery calls from `get_pids_datasets_and_tables` and 146 bigquery calls in `get_table_info_for_dataset`.  This took 0:29:06.475180 minutes to complete.  If this is something we wish to retain and possibly use again, we should try to make this as clean as possible with as few calls to bigquery as possible.

And this time isn't taking into account any actual retraction queries.  It is only the time for running the queries in these two functions for the test project.

amcgrenera-vumc replied:
I was able to simplify this and only call bigquery to get table information with only tables that have a `person_id` column.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,497 @@
+""")
+CHECK_PID_EXIST_QUERY = jinja_env.from_string("""
+SELECT
+COUNT(*) AS count
```
I saw Krishna use `COALESCE(COUNT(*), 0) AS count` and asked him about it.  This will ensure a value is always returned, either 0 or a positive number.  Where `COUNT(*) AS count` could return an empty value.  Then you can do away with line 269:  `return bool(check_pid_df.get_value(0, 'count') > 0)` and replace it with `return check_pid_df.get_value(0, 'count')`.  zero is falsy in python and positive numbers are truthy.

amcgrenera-vumc replied:
Awesome, thank you!

amcgrenera-vumc commented:
@lrwb-aou having some linting errors that i'm looking into, but code is all refactored and ready for you to review again whenever you are ready!

amcgrenera-vumc commented:
@lrwb-aou linting check finally passing-- so this is good to go on my end, whenever you get a chance!

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+                    date_row.project_id, date_row.dataset_id)
+
+                # Create queries based on if the date field is null, if True, will create query based on end_date/start_date
+                if not pd.isnull(date_row.date_column):
```
nit:  This isn't something that has to be changed.  This should work as is.  This is just a comment on the thought process.  When possible, I try to make the conditional logic statements as easy to understand as  I can.  For this one, I might have used `if pd.isnull(date_row.date_column):` then followed that with the queries to build based on end_date/start_date.  And had the `else` statement for the date field query generation.  This removes needing to negate the statement and understand the `if` logic only executes when `isnull` is `False`.

amcgrenera-vumc replied:
you're right I prefer to adjust this anyway, thank you!

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+                        deactivated_pids_dataset=pids_dataset_id,
+                        deactivated_pids_table=pids_table,
+                        date_column=date_row.date_column)
+                    queries_list.append({
+                        clean_consts.QUERY: sandbox_query,
+                        clean_consts.DESTINATION_DATASET: sandbox_dataset,
+                        clean_consts.DESTINATION_TABLE: date_row.table,
+                        'type': 'sandbox'
+                    })
```
You can shorten this section by only setting the `sandbox_query` and `clean_query` in this conditional logid and moving the `queries_list.append` logic outside of the `if/else` logic.  Basically, lines 330-335 are the same as lines 365-370 and lines 345-351 are the same as 381-387.

amcgrenera-vumc replied:
good catch, thank you.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+    query_list = create_queries(args.project_id, args.ticket_number,
+                                args.pids_project_id, args.pids_dataset_id,
+                                args.pids_table)
+    client = bq.get_client(args.project_id)
```
this module uses a lot of queries just to generate the retraction queries.  I would move this before `create_queries` and either pass this client to the `create_queries` function or create the other client creation method listed in a different comment.  Admittedly, I didn't see this first.  I've been following the module dataflow.

amcgrenera-vumc replied:
passing client to `create_queries` I don't think is necessary since it's hitting the `bq.py` to create the client. Which if I refactor `bq.py` this will solve the multiple calls. I am using client in `run_queries` to actually execute the retractions and sandboxing: `response = client.query(query_dict['query'])`

amcgrenera-vumc replied:
I now see what you mean, adjusting this! I will call client once before query_list and pass for all other functions to use.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+    """
+    query_job_ids = []
+    for query_dict in queries:
+        if query_dict['type'] == 'sandbox':
```
nit:  this is safe enough because this isn't being used outside of this module.  But, if you are ever unsure if a python dictionary contains a key, you should use `query_dict.get('key', <default_value>)`.  If the key doesn't exist in the dictionary, it returns the default value you specify (or `None` if you don't specify a default value) instead of raising a `KeyError`.

```
>>> d = {}
>>> s = d.get('alpha', 'a')
>>> print(s)
>>> d['alpha']
```

amcgrenera-vumc replied:
this is good to know, thank you!

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+    for query_dict in queries:
+        if query_dict['type'] == 'sandbox':
+            LOGGER.info(
+                f"Writing rows to be retracted to, using query {query_dict['query']}"
```
nit:  grammar??

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+
+    :param queries: list of queries to run retraction
+    """
+    query_job_ids = []
```
Is this still required?  It's set, but I don't think it's used anymore.?.

amcgrenera-vumc replied:
100% this can be removed. I was copying from nishanth's retraction script but realize this is not needed for mine.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+            query_job_id = response.job_id
+            query_job_ids.append(query_job_id)
+
+    incomplete_jobs = response.exception()
+    if incomplete_jobs is not None:
+        count = len(incomplete_jobs)
+        if incomplete_jobs:
+            LOGGER.info(f"Failed on {count} job ids {incomplete_jobs}")
+            LOGGER.info("Terminating retraction")
+            raise incomplete_jobs
```
your responses are generated in your for loop and you're not saving them to an outside list or variable.  perhaps you want a list to append response objects to.  Then you would loop through the list of response objects to make sure they are finished and finished without exceptions.  You could check with:
```
# assuming response_list is a list of response objects
for resp in response_list:
    if resp.result():
        if resp.exception():
            # logging statements or exception aggregation or your desired custom code
        else:
            # job finished without errors
    else:
        # job isn't finished yet
```

amcgrenera-vumc replied:
I've added `job_result()` after each `client_query()` so it waits for the job to finish. If there is an exception it will append to the `incomplete_jobs` list, if not, it will then hit the `LOGGER.info` stating which table was written with the corresponding query. Please take a look and let me know if you see any adjustments that need to be made.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+            LOGGER.info(
+                f"Truncating table with clean data, using query {query_dict['query']}"
+            )
+            response = client.query(query_dict['query'])
```
You will need to add a [job_config](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client.query) to your query here.  Otherwise, this will select from but will not overwrite the existing table.  This is the [job_config api documentation](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.QueryJobConfig.html#google.cloud.bigquery.job.QueryJobConfig).  Whether you need to add one to line 409 depends on if you need to overwrite an existing table.

amcgrenera-vumc replied:
This has been added, please take a look and let me know if there needs to be any changes. I ran my integration test and checked BQ and it seemed to worked appropriately.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+FROM `{{deactivated_pids_project}}.{{deactivated_pids_dataset}}.{{deactivated_pids_table}}`
+WHERE person_id = {{pid}})
+""")
+CLEAN_QUERY_END_DATE = jinja_env.from_string("""
```
nit:  i think most pycharm users like the double blank lines between high level module functions and statements.

amcgrenera-vumc commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+def get_date_info_for_pids_tables(project_id):
+    """
+    Loop through tables within all datasets and determine if the table has an end_date date or a date field. Filtering
+    out person and death table and keeping only tables with PID and an upload or start/end date associated.
```
This is a good question- I am omitting the person table since the person could still have records before deactivation date, which would in turn need to reference the person table. I then assumed to leave the death table. But I could be completely wrong here. Let me reach out to Caroline.

amcgrenera-vumc replied:
My assumption was wrong, updating this now.

amcgrenera-vumc commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,498 @@
+        "Looping through datasets to filter and create dataframe with correct date field to determine retraction"
+    )
+
+    dataset_obj = bq.list_datasets(project_id)
```
agree, should this be a separate ticket or should I just make this adjustment with this PR?

lrwb-aou replied:
This would be part of this PR, because the example code is assuming it is placed in this module.  You just wouldn't be calling the `bq.list_datasets(project_id)` function anymore.

amcgrenera-vumc replied:
perfect, I will make this adjustment.

amcgrenera-vumc replied:
I was getting an error when trying to change the global variable CLIENT, so I have updated `bq.py` to the following, let me know if that makes sense. I also left in the logic that was there before for `project_id`.

```
CLIENT = None

def get_client(project_id=None):
    """
    Ensure only one client is created and reused

    :param project_id:  project to get a client for
    :returns: a big query client object
    """
    if CLIENT:
        return CLIENT
    if project_id is None:
        LOGGER.info(
            'You should specify project_id for a reliable experience.'
            'Defaulting to %s.', os.environ.get(PROJECT_ID))
        return bigquery.Client()
    return bigquery.Client(project=project_id)
```

lrwb-aou replied:
The above never sets the `CLIENT` global object.  Should it?  My original code example is flawed.  I tested this example more thoroughly.  This will work to define a `CLIENT` variable in the retraction module.
```
from google.cloud import bigquery

# global variable declaration
CLIENT = None

def get_client(project_id):
    """                 
    Ensure only one client is created and reused
                        
    :param project_id:  project to get a client for
    :returns: a big query client object
    """ 
    # declaring that this function should use the global CLIENT variable instead of 
    # creating a local variable.  This local variable was causing the first failure.
    global CLIENT       
    if not CLIENT:
        CLIENT = bigquery.Client(project=project_id)

    return CLIENT
```

amcgrenera-vumc replied:
this has been adjusted per our phone conversation!

amcgrenera-vumc commented:
@lrwb-aou ready for you to take a look again whenever you get the chance.

lrwb-aou commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,182 @@
+        mock_retraction_info.return_value = retraction_info
+
+        # Create and load person_ids and deactivated_date to pid table
+        bq_utils.create_table(self.pid_table_id,
```
Since this is new work, this should be using the bigquery client library instead of `bq_utils`.

amcgrenera-vumc replied:
when trying to convert this to use the bq utility `bq.create_tables`, I found that this function uses the `fields_for(table_name)` in `resources.py`. Am I ok to add a JSON file to the fields folder for this test table? Or should `bq.create_tables` be modified somehow to not rely on those schemas?

lrwb-aou replied:
I absolutely would not add a test table schema to our fields paths.  You are right to consider altering `bq.create_tables`.  You could add an optional `fields` parameter to the signature and pass that value to the `get_table_schema` function.  Again, the `fields` parameter should be optional.  If it exists, it should be used and if it doesn't it should use the `fields_for` function.  Does this make sense?

amcgrenera-vumc replied:
Yes, thank you! I've updated `utils.bq` and tested locally through my integration test. Let me know if any changes need to be made.

lrwb-aou commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,182 @@
+        for cdm_file in test_util.NYC_FIVE_PERSONS_FILES:
+            cdm_file_name = os.path.basename(cdm_file)
+            cdm_table = cdm_file_name.split('.')[0]
+            hpo_table = bq_utils.get_table_id(self.hpo_id, cdm_table)
```
same as above.  should use bigquery client for new work.

amcgrenera-vumc replied:
this function is not in `utils.bq` let me know if I am ok to add? It is just a string concatenation, so I could also add the following.
`hpo_table = self.hpo_id + '_' + cdm_table`

`bq_utils.get_table_id`:
```
def get_table_id(hpo_id, table_name):
    """
    Get the bigquery table id associated with an HPOs CDM table
    :param hpo_id: ID of the HPO
    :param table_name: name of the CDM table
    :return: the table id
    """
    if hpo_id is None:
        return table_name
    # TODO revisit this; currently prefixing table names with hpo_id
    return hpo_id + '_' + table_name
```

lrwb-aou replied:
This might be something to bring up at the next eng_sync meeting.  For now, just leave this line unchanged.  I'm not sure if we should move those into `utils.bq` or somewhere else.

amcgrenera-vumc replied:
sounds good, I'll make sure to add to the agenda for the next eng sync.

lrwb-aou commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,182 @@
+
+TABLE_ROWS_QUERY = 'SELECT * FROM {dataset_id}.__TABLES__ '
+
+EXPECTED_ROWS_QUERY = ('SELECT * '
+                       'FROM {dataset_id}.{table_id} '
+                       'WHERE person_id = {pid} '
+                       'AND {date_column} >= (SELECT deactivated_date '
+                       'FROM {dataset_id}.{pid_table_id} '
+                       'WHERE person_id = {pid})')
```
This query looks like you are checking to make sure data that isn't supposed to be there exists.  That's good.  How about a secondary query to make sure data exists for a participant that should exist.  For example, some data exists after deactivation retraction that should be there.

amcgrenera-vumc commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,519 @@
+                pid = ehr_row.person_id
+
+            # Check if PID is in table
+            if pid is not None and check_pid_exist(pid, date_row, client) > 0:
```
This is good to know, thank you! Always prefer simpler ways.

lrwb-aou commented on data_steward/utils/bq.py:
```diff
@@ -93,7 +109,11 @@ def create_tables(client, project_id, fq_table_names, exists_ok=False):
     failures = []
     for table_name in fq_table_names:
-        schema = get_table_schema(table_name.split('.')[2])
+        if fields:
+            schema = fields
+        else:
+            schema = get_table_schema(table_name.split('.')[2])
+
```
So, this might create a future bug.  It works for the case of building a single table.  I would opt to send the fields to `get_table_schema` to make sure the tables are built with the same schema fields when defined from a json file or from fields provided at run time.
Another concern is applying the same schema to multiple tables in an input table list.  You want to iterate through the fields list in the same fashion as the `fq_table_names`.

amcgrenera-vumc replied:
This has been updated, please take a look.

amcgrenera-vumc commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,519 @@
+
+# Queries to truncate existing tables to remove deactivated EHR PIDS, two different queries for
+# tables with standard entry dates vs. tables with start and end dates
+CLEAN_QUERY_DATE = jinja_env.from_string("""
+SELECT *
+FROM `{{project}}.{{dataset}}.{{table}}`
+WHERE person_id != {{pid}}
+AND {{date_column}} < (SELECT deactivated_date
+FROM `{{deactivated_pids_project}}.{{deactivated_pids_dataset}}.{{deactivated_pids_table}}`
+WHERE person_id = {{pid}})
+""")
+
+CLEAN_QUERY_END_DATE = jinja_env.from_string("""
+SELECT *
+FROM `{{project}}.{{dataset}}.{{table}}`
+WHERE person_id != {{pid}}
+AND (CASE WHEN {{end_date_column}} IS NOT NULL THEN {{end_date_column}} < (SELECT deactivated_date
+FROM `{{deactivated_pids_project}}.{{deactivated_pids_dataset}}.{{deactivated_pids_table}}`
+WHERE person_id = {{pid}} ) ELSE CASE WHEN {{end_date_column}} IS NULL THEN {{start_date_column}} < (
+SELECT deactivated_date
+FROM `{{deactivated_pids_project}}.{{deactivated_pids_dataset}}.{{deactivated_pids_table}}`
+WHERE person_id = {{pid}}) END END)
```
thank you for this find. adjusting both clean queries now. I created datasets too along with the 'end_date_column' to make sure both are working properly. I think adding in the additional query check in the integration test will help confirm as well.

amcgrenera-vumc commented on data_steward/utils/bq.py:
```diff
@@ -61,7 +61,7 @@ def get_table_schema(table_name):
 
 
-def create_tables(client, project_id, fq_table_names, exists_ok=False):
+def create_tables(client, project_id, fq_table_names, exists_ok=False, fields=''):
```
changed to `None`

amcgrenera-vumc commented on tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,226 @@
+from retraction.retract_utils import DEID_REGEX
+
+
+class RetractDataBqTest(unittest.TestCase):
```
updated!

lrwb-aou commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,186 @@
+        row_count_after_retraction = {}
+        for row in results:
+            row_count_after_retraction[row['table_id']] = row['row_count']
+        for table in expected_row_count:
+            self.assertEqual(
+                expected_row_count[table], row_count_before_retraction[table] -
+                row_count_after_retraction[table])
+
+    def tearDown(self):
+        test_util.delete_all_tables(self.bq_dataset_id)
```
@amcgrenera-vumc I'm confused.  I changed the tearDown() function to leave the tables so I could see the changes and I don't understand what I am seeing.  This is the change I made.
```

def tearDown(self):
        print(self.bq_dataset_id)
        #test_util.delete_all_tables(self.bq_dataset_id)
        print("be sure to delete these tables later")
```
When I made this change I ran the test locally and this is my deactivated pid table.  Completely reasonable.
```
person_id | deactivated_date   
2         | 2008-02-01
1234567   | 2019-01-01   
1         | 2007-01-01
```

This is my sandboxed table for the `DCXXX_fake_measurement` table.  It only seems to have sandboxed data for person 1 and not any data for person 2.  Based on the `deactivation_date` for person 2, I would expect all of his/her data to be sandboxed as well.
```
measurement_id | person_id | measurement_date | measurement_datetime
 3             | 1         | 2010-11-05       | 2010-11-05 05:00:00 UTC
 2             | 1         | 2008-09-04       | 2008-09-04 05:00:00 UTC
 1             | 1         | 2008-09-04       | 2008-09-04 05:00:00 UTC 
```

This is the person 1 and 2 data in `fake_measurement` at the end of the test.
```
measurement_id    person_id    measurement_date    measurement_datetime
2                    1               2008-09-04    2008-09-04 05:00:00 UTC
1                    1               2008-09-04    2008-09-04 05:00:00 UTC
3                    1               2010-11-05    2010-11-05 05:00:00 UTC
11                   2               2009-06-02    2009-06-02 05:00:00 UTC
5                    2               2009-06-02    2009-06-02 05:00:00 UTC
6                    2               2009-06-02    2009-06-02 05:00:00 UTC
10                   2               2009-06-02    2009-06-02 05:00:00 UTC
7                    2               2009-06-02    2009-06-02 05:00:00 UTC
9                    2               2009-06-02    2009-06-02 05:00:00 UTC
4                    2               2009-06-02    2009-06-02 05:00:00 UTC
8                    2               2009-06-02    2009-06-02 05:00:00 UTC
13                   2               2010-08-03    2010-08-03 05:00:00 UTC
14                   2               2010-08-03    2010-08-03 05:00:00 UTC
12                   2               2010-08-03    2010-08-03 05:00:00 UTC
17                   2               2009-07-16    2009-07-16 05:00:00 UTC
15                   2               2009-08-11    2009-08-11 05:00:00 UTC
16                   2               2009-01-05    2009-01-05 05:00:00 UTC
18                   2               2010-06-18    2010-06-18 05:00:00 UTC
19                   2               2010-06-18    2010-06-18 05:00:00 UTC
20                   2               2008-10-04    2008-10-04 05:00:00 UTC
22                   2               2009-07-31    2009-07-31 05:00:00 UTC
25                   2               2009-07-31    2009-07-31 05:00:00 UTC
28                   2               2009-07-31    2009-07-31 05:00:00 UTC
21                   2               2009-07-31    2009-07-31 05:00:00 UTC
26                   2               2009-07-31    2009-07-31 05:00:00 UTC
27                   2               2009-07-31    2009-07-31 05:00:00 UTC
23                   2               2009-07-31    2009-07-31 05:00:00 UTC
30                   2               2009-07-31    2009-07-31 05:00:00 UTC
31                   2               2009-07-31    2009-07-31 05:00:00 UTC
24                   2               2009-07-31    2009-07-31 05:00:00 UTC
29                   2               2009-07-31    2009-07-31 05:00:00 UTC
32                   2               2009-04-15    2009-04-15 05:00:00 UTC
33                   2               2008-10-09    2008-10-09 05:00:00 UTC
34                   2               2009-11-11    2009-11-11 05:00:00 UTC
35                   2               2010-02-23    2010-02-23 05:00:00 UTC
```
Correct me if I am wrong.  

1. I expected the table `fake_measurement` to be cleaned of all of `person_id in (1, 2)` data based on their deactivation dates.  Meaning, I didn't expect either individual to have data in the table after running the retraction script on the table.  Is this assumption correct?
2. Why did the sandbox table not contain any data for `person_id = 2`?
3.  If the assumption in question 1 is correct, this test needs to be refactored to ensure it is testing what it should be testing.  It is passing, but **if** the output is wrong, it should be failing.
4.  If the output is correct, please point out what I'm not understanding.

amcgrenera-vumc replied:
I'm looking into this. I was adding the additional query like you had suggested above, which took me down this rabbit hole as well. I'll let you know when this is sorted on my end for you to test again.

amcgrenera-vumc replied:
Just to keep you in the loop the row counts weren't accounting for each individual PID, so it was overwriting to zero since PID 12345 didn't exist in any of the tables with the fake data. I'm adjusting the PIDS table to only include PIDs that exist within that table and making sure it isn't overwriting the counts.

lrwb-aou replied:
@amcgrenera-vumc Thank you.  I felt like I was missing something.  So, the pid I needed to be checking for is 1234567?

amcgrenera-vumc replied:
well my test was wrong, in the loop to get the counts:

```        for query_dict in dropped_row_count_queries:
            response = client.query(query_dict['query'])
            result = response.result()
            expected_dropped_row_count[
                query_dict['destination_table_id']] = result.total_rows
```
It would overwrite the counts for the last loop for that table, so it was overwriting to zero since 123456 didn't exist in the data... if that makes sense. I'm refactoring the code now and will let you know when I think I have it working properly.

amcgrenera-vumc commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,520 @@
+                                args.pids_project_id, args.pids_dataset_id,
+                                args.pids_table)
+    client = get_client(args.project_id)
+    run_queries(query_list, client)
```
that's exactly what i was thinking! and then debugging/running locally to grab that list before running through the process completely.

amcgrenera-vumc replied:
I actually have this already in place and forgot that I did, I created a dataset_list that I appended to when creating the queries. This is output as a logging statement, do you think I should also print this?

```
    LOGGER.info(
        f"Query list complete, retracting ehr deactivated PIDS from the following datasets: "
        f"{dataset_list}")
```

lrwb-aou replied:
No, you don't need to print if you're logging it already.  That should be plenty.

lrwb-aou commented on data_steward/utils/bq.py:
```diff
@@ -93,7 +117,7 @@ def create_tables(client, project_id, fq_table_names, exists_ok=False):
     successes = []
     failures = []
     for table_name in fq_table_names:
-        schema = get_table_schema(table_name.split('.')[2])
+        schema = get_table_schema(table_name.split('.')[2], fields)
```
The changes here are great!  My only concern is if someone sends a list of table names and a single table fields/schema definition to `create_tables`.  I think `fields` may need to be a list of schemas sent to `create_tables`.  Then the `for` loop could be
```
for index, table_name in enumerate(fq_table_names):
    schema = get_table_schema(table_name.split('.')[2], fields[index] if fields else None)
```
FYI:  I haven't tested this.

amcgrenera-vumc replied:
Ahh I see what you're saying. I've made these adjustments and made my list a list of lists of JSON objects. I tested it with my integration test and it will throw an error saying index out of range, if I supply two tables and only one list for the fields/schema.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,520 @@
+                queries_list.append({
+                    clean_consts.QUERY: sandbox_query,
+                    clean_consts.DESTINATION_DATASET: sandbox_dataset,
+                    clean_consts.DESTINATION_TABLE: date_row.table,
```
@amcgrenera-vumc you will probably want to change this to use a `WRITE_APPEND` disposition.

amcgrenera-vumc commented:
@lrwb-aou I've made all the adjustments above and have the retraction working as expected along with the integration test passing, let me know what you think when you get a chance!

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -421,12 +422,16 @@ def run_queries(queries, client):
+        job_config = bigquery.QueryJobConfig(use_query_cache=False,
+                                             destination=query_dict['destination'],
+                                             write_disposition=query_dict['write_disposition'],
+                                             create_disposition='CREATE_IF_NEEDED')
```
Neat!  I hadn't noticed the `create_disposition` feature!

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -391,12 +389,15 @@ def create_queries(project_id, ticket_number, pids_project_id, pids_dataset_id,
                 queries_list.append({
                     clean_consts.QUERY: sandbox_query,
-                    clean_consts.DESTINATION_DATASET: sandbox_dataset,
+                    clean_consts.DESTINATION: date_row.project_id + '.' + date_row.dataset_id + '.' + date_row.table,
```
I think you need to change the `dataset_id` here.  Download this [patch.txt](https://github.com/all-of-us/curation/files/4726263/patch.txt) and apply it.  You can see it is saving the sandboxed rows to the same table as the cleaning queries.  I really don't know how it is doing it.  I was expecting to see sandboxed tables in `lbiggers_ehr_sandbox` but that dataset is empty.  If I'm misinterpreting something, let me know.
```
# Download patch.
# Do the following from  the command line.
# assumes the patch is downloaded to your ~/Downloads folder
# move the patch to your working directory
$ mv ~/Downloads/patch.txt "<path to>/curation/data_steward"

# move to your working directory if you're not already there
$ cd "<path to>/curation/data_steward"

# confirm patch.txt is in your directory
$ ls 

# checkout branch the patch is based on.
$ git checkout am/retract_deactivated_ehr

# apply the patch.    if there have been changes since 2020/06/03 3:00 PM, there may be conflicts.  if there are, I will provide a new patch file.
$ git apply < patch.txt

# verify the patch took
$ git diff

# this should show a change to the integration test as well as some print statements in the actual module
# move up a directory, ensure your environment is sourced, and run the integration test.
$ cd ..
$ ./tests/run_tests.sh -s integration -r retract_deactivated_pids*
```

amcgrenera-vumc replied:
@lrwb-aou This code that you're commenting on is outdated. I downloaded the patch and ran the integration test-- below are my print statements, I also paused the run to verify the sandbox tables were in the correct dataset in BigQuery and they are showing correctly on my end, I have a screenshot below as well. Let me know what you're seeing different?

```
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_condition_occurrence
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_condition_occurrence
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_drug_exposure
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_drug_exposure
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_measurement
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_measurement
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_observation
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_observation
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_procedure_occurrence
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_procedure_occurrence
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_visit_occurrence
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_visit_occurrence
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_condition_occurrence
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_condition_occurrence
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_drug_exposure
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_drug_exposure
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_measurement
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_measurement
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_observation
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_observation
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_procedure_occurrence
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_procedure_occurrence
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_visit_occurrence
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_visit_occurrence
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_condition_occurrence
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_condition_occurrence
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_drug_exposure
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_drug_exposure
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_measurement
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_measurement
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_observation
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_observation
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_procedure_occurrence
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_procedure_occurrence
sandbox destination is:  			aou-res-curation-test.amcgrenera_ehr_sandbox.DCXXX_fake_visit_occurrence
OMOP table destination is:  			aou-res-curation-test.amcgrenera_ehr.fake_visit_occurrence
```

![image](https://user-images.githubusercontent.com/56938070/83693188-3121fd80-a5bb-11ea-92b8-c23842d05dcd.png)

lrwb-aou replied:
This was my fault.  I forgot to fetch the new changes and just did a checkout.  This looks good and correct.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,537 @@
+# Python imports
```
I ran pylint here and got these relevant suggestions.
```
 pylint data_steward/retraction/retract_deactivated_pids.py 
************* Module data_steward.retraction.retract_deactivated_pids
data_steward/retraction/retract_deactivated_pids.py:177:8: W0612: Unused variable 'i' (unused-variable)
data_steward/retraction/retract_deactivated_pids.py:238:12: W0612: Unused variable 'i' (unused-variable)
data_steward/retraction/retract_deactivated_pids.py:17:0: C0412: Imports from package constants are not grouped (ungrouped-imports)
```
When you don't really care about the output you can use an underscore instead, `_`.

lrwb-aou replied:
@amcgrenera-vumc Please group related imports.
```
pylint data_steward/retraction/retract_deactivated_pids.py 
************* Module data_steward.retraction.retract_deactivated_pids
data_steward/retraction/retract_deactivated_pids.py:17:0: C0412: Imports from package constants are not grouped (ungrouped-imports)
```

lrwb-aou commented on tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,226 @@
+# Python imports
```
Running pylint locally gave these relevant suggestions.
```

pylint tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py 
************* Module tests.unit_tests.data_steward.retraction.retract_deactivated_pids_test
tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:60:12: W1505: Using deprecated method assertEquals() (deprecated-method)
tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:50:16: W0612: Unused variable 'i' (unused-variable)
tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:101:16: W0612: Unused variable 'i' (unused-variable)
tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:226:8: W1505: Using deprecated method assertEquals() (deprecated-method)
tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:20:0: W0612: Unused variable 'RetractDataBqTest' (unused-variable)
tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:16:0: C0412: Imports from package constants are not grouped (ungrouped-imports)
tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:17:0: C0412: Imports from package retraction are not grouped (ungrouped-imports)
```
The deprecated alias [is documented here](https://docs.python.org/3.7/library/unittest.html#deprecated-aliases).  I had to look that one up because I wasn't expecting it.

amcgrenera-vumc replied:
thank you for that link!

amcgrenera-vumc commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,275 @@
+from retraction import retract_deactivated_pids
+from io import open
+from constants.cdr_cleaner import clean_cdr as clean_consts
+
```
that makes sense to me, I had other tests where i was told to refactor using `app_identity` to get the project_id, which did scare me a little.

Just to clarify should I alter the following to specifically assign that?

```
self.project_id = 'aou-res-curation-test'
```

lrwb-aou replied:
No, I think getting the app identity the way you are is fine.  I would just add an extra check to make sure it's in the test environment.  

It's a personal preference, but I like the fully qualified table names.  I think it makes it less ambiguous.

amcgrenera-vumc commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,275 @@
+
+    def setUp(self):
+        self.hpo_id = 'fake'
+        self.project_id = app_identity.get_application_id()
```
If I explicitly have `self.project_id = 'aou-res-curation-test'` this will handle that, but let me know if that is bad practice?

lrwb-aou replied:
I don't think I would hard code the test project name here.  I generally don't hard code real values.  But you could use the line "as-is" and add a check that `test` is in the name or raise a runtime error.  Similar to:
```
if 'test' not in self.project_id:
    raise RuntimeError(f"Make sure the project_id is set to test.  project_id is {self.project_id}")
```
Hard coding real values makes it more brittle and exposes some info via the open source nature of the project.

amcgrenera-vumc commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,275 @@
+                             row_count_after_retraction[table])
+
+    def tearDown(self):
+        test_util.delete_all_tables(self.bq_dataset_id)
```
I added a check in case there were other tables in the dataset that were added from a different test, to not delete the sandbox dataset.

amcgrenera-vumc commented:
@lrwb-aou revisions made! Let me know what you think-- checks are failing from the same issue I'm assuming Danyel and Krishna were sorting through.

lrwb-aou commented on tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,226 @@
+        self.mock_bq_client = mock_bq_client_patcher.start()
+        self.addCleanup(mock_bq_client_patcher.stop)
+
+    def test_get_pids_table_info(self):
+        for dataset in self.pids_dataset_list:
+            result_df = self.mock_bq_client.query(
+                retract_deactivated_pids.TABLE_INFORMATION_SCHEMA.render(
+                    project=self.project_id, dataset=dataset)).to_dataframe()
+            expected_pids_tables = []
+            for row in result_df.iterrows():
+                column = getattr(row, 'column_name')
+                table = getattr(row, 'table_name')
+                if 'person_id' in column:
+                    expected_pids_tables.append(table)
+            expected_pids_tables_info_df = result_df[
+                result_df['table_name'].isin(expected_pids_tables)]
+
+            returned_result_df = retract_deactivated_pids.get_pids_table_info(
+                self.project_id, dataset, self.mock_bq_client)
+            self.assertEqual(expected_pids_tables_info_df, returned_result_df)
```
This unit test is only checking that a mocked object equals a mocked object.  So it will always pass.  As it is, it never gets inside the loop at line 50, so none of that ever gets checked.  This test should be rewritten and simplified.

This patch is a unit test that can be incorporated.  It only mocks the return value of the `client.query().to_dataframe()` call and runs the rest of the function.  It sets the expected return value based on the given parameters and asserts a comparison.  It does not rewrite the internals of the function it is testing to generate the expected result.  That test pattern will test that you are getting the same result in both places, not whether the result created by your code is actually correct.

The provided test fails because of the error on line 177 in `retract_deactivated_pids.py`.
[patch.txt](https://github.com/all-of-us/curation/files/4738011/patch.txt)

lrwb-aou commented on tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,226 @@
+                self.project_id, dataset, self.mock_bq_client)
+            self.assertEqual(expected_pids_tables_info_df, returned_result_df)
+
+    def test_get_date_info_for_pids_tables(self):
+
+        date_fields_info_df = pd.DataFrame()
+        dataset_obj = self.mock_bq_client.list_datasets(self.project_id)
+        datasets = [d.dataset_id for d in dataset_obj]
+
+        for dataset in datasets:
+            pids_tables_df = retract_deactivated_pids.get_pids_table_info(
+                self.project_id, dataset, self.mock_bq_client)
+
+            # Keep only records with datatype of 'DATE'
+            date_fields_df = pids_tables_df[pids_tables_df['data_type'] ==
+                                            'DATE']
+
+            # Create df to append to, keeping only one record per table
+            df_to_append = pd.DataFrame(columns=[
+                'project_id', 'dataset_id', 'table', 'date_column',
+                'start_date_column', 'end_date_column'
+            ])
+            df_to_append['project_id'] = date_fields_df['table_catalog']
+            df_to_append['dataset_id'] = date_fields_df['table_schema']
+            df_to_append['table'] = date_fields_df['table_name']
+            df_to_append = df_to_append.drop_duplicates()
+
```
This test has the same problem as the previous test.  This test duplicates the code you are trying to test to create the expected result.  This test should also fail, but it doesn't because it contains the same errors.  Line 101 is incorrect and the test should fail.

The assertion on line 119 passes because both the expected and result data frames are empty.  I didn't write a patch file for this, but I can help if needed.

lrwb-aou commented on tests/unit_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,226 @@
+
+    @mock.patch('utils.bq.query')
+    @mock.patch(
+        'retraction.retract_deactivated_pids.get_date_info_for_pids_tables')
```
This unit test is not providing any worthwhile information.  It ends by comparing two blank lists.

lrwb-aou commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,283 @@
+from io import open
+
+# Third party imports
+import app_identity
```
```
pylint tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py 
************* Module tests.integration_tests.data_steward.retraction.retract_deactivated_pids_test
tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:9:0: C0411: third party import "import pandas as pd" should be placed before "import app_identity" (wrong-import-order)
tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:10:0: C0411: third party import "import mock" should be placed before "import app_identity" (wrong-import-order)
```
Please fix the import orders.

lrwb-aou commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,283 @@
+
+    @mock.patch(
+        'retraction.retract_deactivated_pids.get_date_info_for_pids_tables')
+    def test_integration_queries_to_retract_from_fake_dataset(
```
This test passes, but should fail due to logic errors.  It passes because the errors are in `get_date_info_for_pids_tables`, which is mocked, and in functions called by `get_date_info_for_pids_tables`.  Instead of mocking the entire function, mock the calls to the external resources only and set their return values as they are expected to look.  See the comment in `retract_deactivated_pids.py` for more info on the error and the comments in the unit tests to understand why this error was logically missed.

amcgrenera-vumc commented:
@lrwb-aou I have updated the unit tests and they are passing. Please take a look when you get a chance. I also fixed a bug in the integration test.

lrwb-aou commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,283 @@
+                None, 'visit_end_date'
+            ]
+        }
+        retraction_info = pd.DataFrame(data=d)
+        mock_retraction_info.return_value = retraction_info
```
Nice use of the mock return value to set up your test.

lrwb-aou commented on tests/integration_tests/data_steward/retraction/retract_deactivated_pids_test.py:
```diff
@@ -0,0 +1,283 @@
+    def tearDown(self):
+        test_util.delete_all_tables(self.bq_dataset_id)
+        test_util.delete_all_tables(self.bq_sandbox_dataset_id)
+        tables = list(self.client.list_tables(self.bq_sandbox_dataset_id))
+        if not tables:
+            self.client.delete_dataset(self.bq_sandbox_dataset_id)
```
This check is not strictly necessary.  The test_util is going to delete every table in the sandbox dataset when you call it.  You should be fine to just remove the sandbox dataset after that.  If a test left data in there, it shouldn't have.

amcgrenera-vumc replied:
Ok great, I will remove this check, thanks!

amcgrenera-vumc commented:
@lrwb-aou I added in a couple more logging statements and also a check to remove synthetic data, curation sandbox and vocabulary datasets (to try and help reduce timing of run). Caroline has approved the datasets so I will run the retraction and if you are good with these additions I will merge this.

amcgrenera-vumc commented:
@lrwb-aou One more change-- I adjusted `check_pid_exist` and updated the query to check not only is the PID in the table, but that the date is past the deactivated date. This will help cut back on time because it will not be creating unnecessary queries.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,608 @@
+WHERE person_id = {{pid}}
+""")
+
+CHECK_PID_EXIST_DATE_QUERY = jinja_env.from_string("""
+SELECT
+COALESCE(COUNT(*), 0) AS count
+FROM `{{project}}.{{dataset}}.{{table}}`
+WHERE person_id IN (SELECT person_id
+FROM `{{deactivated_pids_project}}.{{deactivated_pids_dataset}}.{{deactivated_pids_table}}`)
+AND {{date_column}} > (SELECT MIN(deactivated_date)
+FROM `{{deactivated_pids_project}}.{{deactivated_pids_dataset}}.{{deactivated_pids_table}}`)
+""")
```
So, this is checking that a person exists in the deactivated table and a date field exists that is greater then the minimum deactivated date in the deactivation table.  Even if the minimum deactivation date belongs to a different person id.  Is this what you intend?

amcgrenera-vumc replied:
Correct! It's a generalized select to try and eliminate tables that queries need to be created for. If I went through every deactivated date per PID it would be a ton more queries and take a lot longer.

lrwb-aou commented on data_steward/retraction/retract_deactivated_pids.py:
```diff
@@ -0,0 +1,608 @@
+    datasets = [d.dataset_id for d in dataset_obj]
+
+    # Remove synthetic data, vocabulary, curation sandbox and previous naming convention datasets
+    prefixes = ('SR', 'vocabulary', 'curation', 'combined', '2018', 'R2018',
+                'rdr')
+    datasets = [x for x in datasets if not x.startswith(prefixes)]
```
neat!  I've never used tuples with the `startswith` method before.

amcgrenera-vumc commented:
Thanks @nishanthpp93 ! I agree that reducing the amount of queries would definitely be beneficial, but it would take a good amount of refactoring to adjust. For now I am going to keep this as is so we can finally run the retraction. Lauren and I had discussed previously this module will most likely need updating if we are going to use it multiple times to shorten the run time, so it would be good to also adjust the query then.
